helper_crates/document-features/lib.rs

```diff
@@ -0,0 +1,327 @@
+
+    let result = process_toml(&cargo_toml).map_err(|e| error(&e))?;
+    Ok(std::iter::once(proc_macro::TokenTree::from(proc_macro::Literal::string(&result))).collect())
+}
+
+fn process_toml(cargo_toml: &str) -> Result<String, String> {
+    // Get all lines between the "[features]" and the next block
+    let lines = cargo_toml
+        .lines()
+        .skip_while(|l| l.trim() != "[features]")
+        .skip(1) // skip the [features] line
+        .map(str::trim)
+        .take_while(|l| !l.starts_with("["))
+        // and skip empty lines and comments that are not docs comments
+        .filter(|l| {
+            !l.is_empty() && (!l.starts_with("#") || l.starts_with("##") || l.starts_with("#!"))
+        });
+    let mut top_comment = String::new();
+    let mut current_comment = String::new();
+    let mut features = vec![];
+    let mut default_features = HashSet::new();
+    for line in lines {
+        if let Some(x) = line.strip_prefix("#!") {
+            if !x.is_empty() && !x.starts_with(" ") {
+                continue; // it's not a doc comment
+            }
+            if !current_comment.is_empty() {
+                return Err("Cannot mix ## and #! comments between features.".into());
+            }
+            writeln!(top_comment, "{}", x).unwrap();
+        } else if let Some(x) = line.strip_prefix("##") {
+            if !x.is_empty() && !x.starts_with(" ") {
+                continue; // it's not a doc comment
+            }
+            writeln!(current_comment, "{}", x).unwrap();
+        } else if let Some((dep, rest)) = line.split_once("=") {
+            if dep.trim() == "default" {
+                let defaults = rest
+                    .trim()
+                    .strip_prefix("[")
+                    .and_then(|r| r.strip_suffix("]"))
+                    .ok_or_else(|| format!("Parse error while parsing dependency {}", dep))?
+                    .split(",")
+                    .map(|d| d.trim().trim_matches(|c| c == '"' || c == '\'').trim())
+                    .filter(|d| !d.is_empty());
+                default_features.extend(defaults);
+            } else {
+                // Do not show features that do not have documentation
+                if !current_comment.is_empty() {
+                    features.push((
+                        dep.trim(),
+                        std::mem::take(&mut top_comment),
+                        std::mem::take(&mut current_comment),
+                    ));
+                }
+            }
+        } else {
+            return Err(format!("Parse error while processing the line:\n{}", line));
+        }
+    }
+    if !current_comment.is_empty() {
+        return Err("Found comment not associated with a feature".into());
+    }
+    if features.is_empty() {
+        return Err("Could not find features in Cargo.toml".into());
+    }
+    let mut result = String::from("## Feature flags\n\n");
```

Comment:
I'm wondering if this is actually a good idea to have this title there.
Some crate uses `#` while some other uses `##` depending on how they nest sections.
Maybe i can just leave that out and one must write it in lib.rs or in Cargo.toml as a `#!` comment

Reply:
Yes, makes sense to let the using crate define that.